### PR TITLE
[Xamarin.Android.Build.Tasks] fix for empty $(AndroidNdkDirectory)

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: https://github.com/xamarin/xamarin-android/pull/3402

When reviewing #3402, we found two things that seem wrong™:

* A case where `AndroidNdkDirectory` in the
  `<GetAdditionalResourcesFromAssemblies/>` MSBuild task, was
  `[Required]`. How did that ever work?
* In cases where `$(AndroidNdkDirectory)` is blank, it appears to get
  set to `\`. Ok that's how it worked...

The first fix was to just remove `[Required]`, I did not find any
other places this was incorrect. I also had to adjust
`<GetAdditionalResourcesFromAssemblies/>` to account for the value
being blank.

Then I revisited the pattern we have in the `_ResolveMonoAndroidSdks`
MSBuild target:

    <CreateProperty Value="$(AppSettingsDirectory)">
        <Output TaskParameter="Value" PropertyName="_AppSettingsDirectory"/>
    </CreateProperty>
    <CreateProperty Value="$(_AppSettingsDirectory)\">
        <Output TaskParameter="Value" PropertyName="_AppSettingsDirectory"
            Condition="!HasTrailingSlash('$(_AppSettingsDirectory)')" />
    </CreateProperty>

That is a lot of logic, which actually doesn't work properly if the
incoming value is blank... The end result is `\`.

This would be more properly written as:

    <_AppSettingsDirectory>$([MSBuild]::EnsureTrailingSlash($(AppSettingsDirectory)))</_AppSettingsDirectory>

https://docs.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2019#msbuild-property-functions

Using `$([MSBuild]::EnsureTrailingSlash())` does the following:

    If the given path doesn't have a trailing slash then add one. If the path is an empty string, does not modify it.

I have reworked the `_ResolveMonoAndroidSdks` target to use this
simpler pattern instead.

I also wrote a few tests around this target, which runs the
`_ResolveMonoAndroidSdks` target directly. Each test takes < 1 second
to complete, since it isn't actually doing a full build.

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 1]
```diff
@@ -922,55 +922,17 @@ because xbuild doesn't support framework reference assemblies.
...
-
-	<CreateProperty Value="$(AndroidSdkBuildToolsBinPath)\">
-		<Output TaskParameter="Value" PropertyName="AndroidSdkBuildToolsBinPath"
-				Condition="!HasTrailingSlash('$(AndroidSdkBuildToolsBinPath)')"
-		/>
-	</CreateProperty>
+	<PropertyGroup>
+		<_AppSettingsDirectory>$([MSBuild]::EnsureTrailingSlash($(AppSettingsDirectory)))</_AppSettingsDirectory>
+		<_ApkDebugKeyStore Condition=" '$(_ApkDebugKeyStore)' == '' ">$(_AppSettingsDirectory)debug.keystore</_ApkDebugKeyStore>
+		<_MonoAndroidToolsDirectory>$([MSBuild]::EnsureTrailingSlash($(MonoAndroidToolsDirectory)))</_MonoAndroidToolsDirectory>
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets]: I believe this should be:

```
$([MSBuild]::EnsureTrailingSlash($(_MonoAndroidToolsDirectory)))
```

Note `_` in `_MonoAndroidToolsDirectory`.

Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets]: This one actually looks right, the incoming value from our existing code doesn't have the `_`: https://github.com/xamarin/xamarin-android/blob/5695ecb0958de9cae81e469d5b890ef2745a74bf/src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets#L939-L941

I read through the diff again to double-check... I think they are the same as what we had before?

**Comment by kzu** [comment]
When this is merged, we can start exploring setting up the A/B experiment to enable background builds :)

File: src/Xamarin.Android.Build.Tasks/Tasks/GetAdditionalResourcesFromAssemblies.cs [hunk 2]
```diff
@@ -71,8 +70,14 @@ string SubstituteEnvVariables (string path)
 			MatchEvaluator replaceEnvVar = (Match m) => {
 				var e = m.Groups [2].Value;
 				switch (e) {
-				case "ANDROID_SDK_PATH": return AndroidSdkDirectory.TrimEnd (Path.DirectorySeparatorChar);
-				case "ANDROID_NDK_PATH": return AndroidNdkDirectory.TrimEnd (Path.DirectorySeparatorChar);
+				case "ANDROID_SDK_PATH":
+					return AndroidSdkDirectory.TrimEnd (Path.DirectorySeparatorChar);
+				case "ANDROID_NDK_PATH":
+					//NOTE: AndroidNdkDirectory is not [Required]
```
Comment (dellis1972) [inline src/Xamarin.Android.Build.Tasks/Tasks/GetAdditionalResourcesFromAssemblies.cs]: @jonathanpeppers this code path is only used for `Android.IncludeAndroidResourcesFromAttribute`, `Android.NativeLibraryReferenceAttribute` and `Java.Interop.JavaLibraryReferenceAttribute` entries which are normally included in one of the Support Libraries. So we might have an attribute like

```
[assembly: Android.NativeLibraryReference ("ANDROID_NDK_PATH/platforms/android-29/arch-arm64/usr/lib/libEGL.so")]
```

 Now I'm not sure if there are any of our support libraries which do use that but if there are and `AndroidNdkDirectory` is blank we might have a problem with the build later. We should probably add a unit test to make sure this works and/or raises a decent error message if `AndroidNdkDirectory` is blank and we hit one of these.

**Comment by brendanzagaeski** [comment]
A new preview version has now been published that includes a fix for this item.

Fix included in Xamarin.Android 10.0.0.40.

Fix included on Windows in Visual Studio 2019 version 16.3 Preview 3. To try the preview version that includes the fix, [check for the latest updates](https://docs.microsoft.com/visualstudio/install/update-visual-studio?view=vs-2019) in [Visual Studio Preview](https://visualstudio.microsoft.com/vs/preview/).

Fix included on macOS in Visual Studio 2019 for Mac version 8.3 Preview 3. To try the preview version that includes the fix, check for the latest updates on the **Preview** [updater channel](https://docs.microsoft.com/visualstudio/mac/update).

**Comment by brendanzagaeski** [comment]
*Release status update*

A new Release version has now been published that includes the fix for this item.

Fix included in Xamarin.Android 10.0.0.43

Fix included on Windows in Visual Studio 2019 version 16.3. To get the new version that includes the fix, [check for the latest updates](https://docs.microsoft.com/visualstudio/install/update-visual-studio?view=vs-2019) or install the latest version from <https://visualstudio.microsoft.com/downloads/>.

Fix included on macOS in Visual Studio 2019 for Mac version 8.3. To get the new version that includes the fix, check for the latest updates on the **Stable** [updater channel](https://docs.microsoft.com/visualstudio/mac/update).

**Pull request merged**